gateway-js/src/supergraphManagers/LegacyFetcher/index.ts

```diff
@@ -0,0 +1,263 @@
+      result.serviceDefinitions,
+    );
+
+    if (!supergraphSdl) {
+      throw new Error(
+        "A valid schema couldn't be composed. Falling back to previous schema.",
+      );
+    } else {
+      this.schema = schema;
+      return supergraphSdl;
+    }
+  }
+
+  private createSchemaFromServiceList(serviceList: ServiceDefinition[]) {
+    this.config.logger?.debug(
+      `Composing schema from service list: \n${serviceList
+        .map(({ name, url }) => `  ${url || 'local'}: ${name}`)
+        .join('\n')}`,
+    );
+
+    const compositionResult = composeAndValidate(serviceList);
+
+    if (compositionHasErrors(compositionResult)) {
+      const { errors } = compositionResult;
+      throw Error(
+        "A valid schema couldn't be composed. The following composition errors were found:\n" +
+          errors.map((e) => '\t' + e.message).join('\n'),
+      );
+    } else {
+      const { supergraphSdl } = compositionResult;
+      for (const service of serviceList) {
+        this.getDataSource?.(service);
+      }
+
+      const schema = buildComposedSchema(parse(supergraphSdl));
+
+      this.config.logger?.debug('Schema loaded and ready for execution');
+
+      // This is a workaround for automatic wrapping of all fields, which Apollo
+      // Server does in the case of implementing resolver wrapping for plugins.
+      // Here we wrap all fields with support for resolving aliases as part of the
+      // root value which happens because aliases are resolved by sub services and
+      // the shape of the root value already contains the aliased fields as
+      // responseNames
+      return {
+        schema: wrapSchemaWithAliasResolver(schema),
+        supergraphSdl,
+      };
+    }
+  }
+
+  private beginPolling() {
+    this.state = { phase: 'polling' };
+    this.poll();
+  }
+
+  private poll() {
+    this.timerRef = setTimeout(async () => {
+      if (this.state.phase === 'polling') {
+        const pollingPromise = resolvable();
+
+        this.state.pollingPromise = pollingPromise;
+        try {
+          const maybeNewSupergraphSdl = await this.updateSupergraphSdl();
+          if (maybeNewSupergraphSdl) {
+            this.update?.(maybeNewSupergraphSdl);
+          }
+        } catch (e) {
+          this.logUpdateFailure(e);
+        }
+        pollingPromise.resolve();
+      }
+
+      this.poll();
+    }, this.config.pollIntervalInMs!);
+  }
+
+  private logUpdateFailure(e: any) {
+    this.config.logger?.error(
+      'UplinkFetcher failed to update supergraph with the following error: ' +
+        (e.message ?? e),
+    );
+  }
+}
+
+// We can't use transformSchema here because the extension data for query
+// planning would be lost. Instead we set a resolver for each field
+// in order to counteract GraphQLExtensions preventing a defaultFieldResolver
+// from doing the same job
+function wrapSchemaWithAliasResolver(schema: GraphQLSchema): GraphQLSchema {
```

Comment:
Why does LegacyFetcher and LocalCompose need to do this but IntrospectAndCompose doesn't?

Also why is this function copy-and-pasted three times?

Reply:
I never quite understood why this thing was necessary but I *think* in the current PR, gateway/index.ts now does almost everything through a single simple codepath which already runs this, so the individual managers don't need to do it again.

Reply:
d49dcf9b